website/versioned_docs/version-5.x/api.md

```diff
@@ -638,100 +638,107 @@ Sets the global configuration for unload timeouts.
+### Custom Events
 
-<!-- TODO: are these events augmented like the addErrorHandler Error is? -->
+single-spa fires a series of [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent) whenever it reroutes. A reroute occurs whenever the browser URL changes in any way or a `triggerAppChange` is called. The custom events are fired to the `window`. Each custom event has a [`detail` property](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/detail) with the following properties:
 
 ```js
 window.addEventListener('single-spa:before-routing-event', evt => {
-  const originalEvent = evt.detail;
-  console.log('single-spa event', originalEvent);
+	const { originalEvent, newAppStatuses, appsByNewStatus, totalAppChanges } = evt.detail;
+	console.log('original event that triggered this single-spa event', originalEvent); // PopStateEvent | HashChangeEvent | undefined
+	console.log('the new status for all applications after the reroute finishes', newAppStatuses) // { app1: MOUNTED, app2: NOT_MOUNTED }
+	console.log('the applications that changed, grouped by their status', appsByNewStatus) // { MOUNTED: ['app1'], NOT_MOUNTED: ['app2'] }
+	console.log('number of applications that changed status so far during this reroute', totalAppChanges); // 2
 })
 ```
 
-## before routing event
-
-```js
-window.addEventListener('single-spa:before-routing-event', () => {
-	console.log('single-spa is about to mount/unmount applications!');
-});
-```
+The following table shows the order in which the custom events are fired during a reroute:
 
-A `single-spa:before-routing-event` event is fired before every routing event occurs, which is after each hashchange, popstate, or triggerAppChange, even if no changes to registered applications were necessary.
+| Event order | Event Name | Condition for firing |
+| ----------- | ---------- | -------------------- |
+| 1 | `single-spa:before-app-change` or `single-spa:before-no-app-change` | Will any applications change status? |
+| 2 | `single-spa:before-routing-event` | &mdash; |
+| 3 | `single-spa:before-mount-routing-event` | &mdash; |
+| 4 | `single-spa:before-first-mount` | Is this the first time any application is mounting? |
+| 5 | `single-spa:first-mount` | Is this the first time any application was mounted? |
+| 5 | `single-spa:app-change` or `single-spa:no-app-change` | Did any applications change status? |
+| 6 | `single-spa:routing-event` | &mdash; |
 
-## before mount routing event
+### before-app-change event
```

Comment:
I reordered these events to be in the order in which they are fired during reroute.